hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventResponseConverter.java

```diff
@@ -54,18 +54,37 @@
       LogEventsResponse logEventsResponse) {
     return Observable.fromIterable(logEventsResponse.getLogEventsList())
         .concatMapSingle(
-            logEventsResponseVar -> this.convert(attributeRequests, logEventsResponseVar))
-        .collect(Collectors.groupingBy(logEvent -> (String) logEvent.attribute(foreignIdAttribute)))
+            logEventsResponseVar ->
+                this.convert(foreignIdAttribute, attributeRequests, logEventsResponseVar))
+        .collect(
+            Collectors.groupingBy(
+                spanLogEventPair -> spanLogEventPair.spanId,
+                Collectors.mapping(v -> v.logEvent, Collectors.toList())))
         .map(
             spanIdVsLogEventsMap -> new SpanLogEventsResponse(spansResponse, spanIdVsLogEventsMap));
   }
 
-  private Single<LogEvent> convert(
+  private Single<SpanLogEventPair> convert(
+      String foreignIdAttribute,
       Collection<AttributeRequest> request,
       org.hypertrace.gateway.service.v1.log.events.LogEvent logEvent) {
     return this.attributeMapConverter
         .convert(request, logEvent.getAttributesMap())
-        .map(ConvertedLogEvent::new);
+        .map(
+            attributeMap ->
+                new SpanLogEventPair(
+                    logEvent.getAttributesMap().get(foreignIdAttribute).getString(),
+                    new ConvertedLogEvent(attributeMap)));
+  }
+
+  static class SpanLogEventPair {
```

Comment:
Can we have it as a private static class,  with lombok annotations, so that you don't have to specify a constructor?